### PR TITLE
feat(ecr-assets): Support docker outputs flag

### LINKED ISSUE #20566: assets: allow passing --output flag to Docker build
### Describe the feature

[`AssetImageProps`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AssetImageProps.html) should allow setting the `docker build` [`--output`](https://docs.docker.com/engine/reference/commandline/buildx_build/#output) flag.

### Use Case

Specifically, we need to set `--output=type=docker`. It could also be expressed by the shorthand [`--load` flag](https://docs.docker.com/engine/reference/commandline/buildx_build/#load), but I would prefer a solution which allows the user to pass any `--output`, to avoid having to adapt CDK every time Docker adds new output options.

#### Background
We use CDK to build and deploy an ARM64 image on an x86-based (i.e. AMD64) Linux EC2 instance. Only Docker desktop offers multi-platform builds out of the box, so we had to install it, as per [Docker's guide](https://docs.docker.com/buildx/working-with-buildx/#set-buildx-as-the-default-builder). We make `buildx` the default builder by running `docker buildx install` and then configuring a builder for `arm64`. Invoking `docker build ...` (like CDK does) is then actually re-routed to use `buildx` instead of the legacy Docker builder.

BUT `buildx` [by default doesn't publish the built image into the local registry](https://docs.docker.com/engine/reference/commandline/buildx_build/#output) and requires an additional `--output` flag (or a shorthand like `--load`) to actually store and use the image that was built. This is reflected by warnings shown by `buildx` when invoking `docker build` without setting the `--output` flag (or any of its shorthands like `--push` or `--load`:

```
No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
```

Because CDK `AssetImageProps` builder API doesn't allow us to set the output for the `docker build` step, it fails after building because it tries to tag an image which doesn't exist:

```
docker tag cdkasset-da3d3b814c3ba23b82b05bcac15892e50aa87059a01ba3dbada59dfef2369aaf foo.dkr.ecr.eu-west-1.amazonaws.com/cdk-hnb659fds-container-assets-foo-eu-west-1:da3d3b814c3ba23b82b05bcac15892e50aa87059a01ba3dbada59dfef2369aaf

Error response from daemon: No such image: cdkasset-da3d3b814c3ba23b82b05bcac15892e50aa87059a01ba3dbada59dfef2369aaf:latest
```

Workarounds like @[nathanpeck's](https://github.com/nathanpeck) https://github.com/aws/aws-cdk/issues/12472#issuecomment-904937818 are not an option for us, since we need to be able to synthesize our CDK stack without building all the Docker images (which requires long compilation etc).

Our current workaround is to rename the `docker` binary to `docker_real` and instead create a `docker` bash script on the path which adds the flag:

```sh
#!/usr/bin/env bash
if [ "$1" = "build" ]; then
  echo "Adding --load param to docker build command"
  docker_real build --load "${@:2}"
else
  docker_real "$@"
fi
```

But this is obviously a hack and the proper solution would be for CDK `AssetImageProps` builder to allow us to add the `--output` flag (or preferably: any flag) to its invocation of `docker build`.

### Proposed Solution

The [`AssetImageProps`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AssetImageProps.html) builder should get a new method `output`, which works similar to [the existing`buildArgs` method](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AssetImageProps.html#buildargs).

For example, calling this new `output` method with a map like `{"type": "local", "dest": "path"}` would result in `--output=type=local&dest=path`.

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.26.0

### Environment details (OS name and version, etc.)

Amazon Linux 2

### PR DESCRIPTION
**Pull request by jesse-peters**
This adds the `--output` flag as an option when building docker containers. This fixes #20566.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Construct Runtime Dependencies:

* [ ] This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-construct-runtime-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jesse-peters** [comment]
I think this should be ready for a review now? Happy to make any changes needed though!

**Comment by jesse-peters** [comment]
Hi, this has been open for a few weeks (over the holidays I know) - I was just wondering if anyone might be able to provide feedback or help clear up the ci warning.

**Comment by jesse-peters** [comment]
Hello, is there anything I can do to make sure this PR doesnt get cleaned up by the bot as Im about to hit 4 weeks? 
@aws-cdk-automation

**Comment by jesse-peters** [comment]
Hey @mrgrain - sorry to bug you with a ping. I saw you've worked on some similar PRs on this repo, and I was hoping to get some feedback before the bot closes this PR. I'm sure you're pretty busy getting past the holidays, I just don't want to lose momentum here.

**Comment by mrgrain** [comment]
All good @jesse-peters . I'm not back from annual leave yet, but assigned the PR to me. If it gets auto closed before I get to it, just reopen a new one and tag me .

**Comment by jesse-peters** [comment]
Sounds great, thanks!

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts [hunk 1]
```diff
@@ -47,7 +47,7 @@ export class NetworkMode {
   /**
    * @param mode The networking mode to use for docker build
    */
-  private constructor(public readonly mode: string) {}
+  private constructor(public readonly mode: string) { }
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:50]: Just to confirm: Did these come from the linter? If so - great! Otherwise I'd ask you to keep unrelated formatting changes to a minimum.

File: packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts [hunk 2]
```diff
@@ -80,6 +80,14 @@ export interface DockerImageSource {
    * @default - current machine platform
    */
   readonly platform?: string;
+
+  /**
+   * Outputs
+   *
+   * @default - No outputs
```
Comment (mrgrain) [inline packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts:87]: Let's be a bit more descriptive here. It's also okay to link to the docker docs with `@see`

File: packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts [hunk 3]
```diff
@@ -80,6 +80,14 @@ export interface DockerImageSource {
...
+   */
+  readonly dockerOutputs?: string[];
+
```
Comment (mrgrain) [inline packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts:90]:
```suggestion
```

File: packages/@aws-cdk/cloud-assembly-schema/lib/cloud-assembly/metadata-schema.ts [hunk 4]
```diff
@@ -145,6 +145,14 @@ export interface ContainerImageAssetMetadataEntry extends BaseAssetMetadataEntry
...
    */
   readonly platform?: string;
+
+  /**
+   * Outputs to pass to the `docker build` command
+   *
+   * @default no outputs are passed
+   */
+  readonly outputs?: string[];
+
```
Comment (mrgrain) [inline packages/@aws-cdk/cloud-assembly-schema/lib/cloud-assembly/metadata-schema.ts:155]:
```suggestion
```

File: packages/@aws-cdk/cloud-assembly-schema/schema/assets.schema.json [hunk 5]
```diff
@@ -162,6 +162,13 @@
                 "platform": {
                     "description": "Platform to build for. _Requires Docker Buildx_.\n\nSpecify this property to build images on a specific platform/architecture. (Default - current machine platform)",
                     "type": "string"
+                },
+                "dockerOutputs": {
+                    "description": "Outputs (Default - No outputs)",
```
Comment (mrgrain) [inline packages/@aws-cdk/cloud-assembly-schema/schema/assets.schema.json:167]: Needs better description (I know these are repetitive)

Comment (mrgrain) [reply packages/@aws-cdk/cloud-assembly-schema/schema/assets.schema.json:167]: FIY: This file might actually be generated....

File: packages/@aws-cdk/cx-api/lib/assets.ts [hunk 6]
```diff
@@ -15,7 +15,7 @@ export const ASSET_RESOURCE_METADATA_DOCKER_BUILD_ARGS_KEY = 'aws:asset:docker-b
 export const ASSET_RESOURCE_METADATA_DOCKER_BUILD_TARGET_KEY = 'aws:asset:docker-build-target';
 export const ASSET_RESOURCE_METADATA_PROPERTY_KEY = 'aws:asset:property';
 export const ASSET_RESOURCE_METADATA_IS_BUNDLED_KEY = 'aws:asset:is-bundled';
-
+export const ASSET_RESOURCE_METADATA_DOCKER_OUTPUTS_KEY = 'aws:asset:docker-outputs';
```
Comment (mrgrain) [inline packages/@aws-cdk/cx-api/lib/assets.ts:18]:
```suggestion
export const ASSET_RESOURCE_METADATA_DOCKER_OUTPUTS_KEY = 'aws:asset:docker-outputs';

```

**Review by mrgrain — CHANGES_REQUESTED** [review]
Code looks good. Just some very minor fixes. I also need to run the CLI tests then which might take a couple days for me to get to.

---

More generally speaking - would there be a way to detect if `buildx` is running and default the ouput to `--output=type=docker` in that case? Or maybe we really should have an image asset that explicitly uses buildx. 🤔

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts (hunk 1 above)
Comment (jesse-peters) [reply packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:50]: Yep, this came from the linter.

File: packages/@aws-cdk/cloud-assembly-schema/schema/assets.schema.json (hunk 5 above)
Comment (jesse-peters) [reply packages/@aws-cdk/cloud-assembly-schema/schema/assets.schema.json:167]: Yep, it is. Thanks!

**Comment by mrgrain** [comment]
> I think an image asset that's built for buildx would probably be the best in the long run. This flag exists on standard docker as well, but there are a number of new ones that are buildx specific.

Interesting idea. I like it. It might require a lot of duplicated constructs and flags in various places. 🤔

File: packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts (hunk 2 above)
Comment (mrgrain) [reply packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts:87]: @jesse-peters Once you've update this doc block and re-generated the schema, the PR should be good to go.

**Comment by jesse-peters** [comment]
Yea, there probably would be a fair bit of duplication between the two.

I think I hit the issues you pointed out on this PR. But I'll keep mulling over a new asset/how we might modify the current one to play nicer with `buildx`.

File: packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts (hunk 2 above)
Comment (jesse-peters) [reply packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts:87]: @mrgrain I believe I pushed this last night?

Comment (mrgrain) [reply packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts:87]: Yep I saw that. This is a different place that you've probably just missed. Unfortunately there's a lot of repetition with this. =/

Comment (jesse-peters) [reply packages/@aws-cdk/cloud-assembly-schema/lib/assets/docker-image-asset.ts:87]: Oops, you're right! I just pushed an update.

**Review by mrgrain — APPROVED** [review]
🚀
